mlflow/models/utils.py

```diff
@@ -142,28 +142,20 @@ def _handle_dataframe_nans(df: pd.DataFrame):
 
         def _coerce_to_pandas_df(input_ex):
             if isinstance(input_ex, dict):
-                if all(_is_scalar(x) for x in input_ex.values()):
-                    input_ex = pd.DataFrame([input_ex])
-                elif all(isinstance(x, (str, list)) for x in input_ex.values()):
-                    for value in input_ex.values():
-                        if isinstance(value, list) and not all(_is_scalar(x) for x in value):
-                            raise TypeError(
-                                "List values within dictionaries must be of scalar type."
-                            )
-                    input_ex = pd.DataFrame(input_ex)
-                else:
-                    raise TypeError(
-                        "Data in the dictionary must be scalar or of type numpy.ndarray"
+                # We need to be compatible with infer_schema's behavior, where
+                # it infers each value's type directly.
+                if all(
+                    isinstance(x, str) or (isinstance(x, list) and all(_is_scalar(y) for y in x))
+                    for x in input_ex.values()
+                ):
+                    _logger.warning(
+                        "We convert dictionary to pandas DataFrame by assuming each key "
+                        "is a column, and in total one row of data. If you would like to "
+                        "save data as multiple rows, please convert your data to pandas "
+                        "DataFrame before passing to input_example."
                     )
-            elif isinstance(input_ex, list):
-                for i, x in enumerate(input_ex):
-                    if isinstance(x, np.ndarray) and len(x.shape) > 1:
-                        raise TensorsNotSupportedException(f"Row '{i}' has shape {x.shape}")
-                if all(_is_scalar(x) for x in input_ex):
-                    input_ex = pd.DataFrame([input_ex], columns=range(len(input_ex)))
-                else:
-                    input_ex = pd.DataFrame(input_ex)
-            elif isinstance(input_ex, (str, bytes)):
+                input_ex = pd.DataFrame([input_ex])
+            elif np.isscalar(input_ex):
                 input_ex = pd.DataFrame([input_ex])
```

Comment:
These two conditions result in the same transformation. Do we want to keep the logic separator here for readability or is that unintentional? (I'm fine with keeping them separate to make this more readable as this is a very large logical chain and historically this function has been very hard to read; any simplification is a bonus).

Reply:
I'll keep it :D